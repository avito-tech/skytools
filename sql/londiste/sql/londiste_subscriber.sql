
set client_min_messages = 'warning';
\set VERBOSITY 'terse'

create table testdata (
    id serial primary key,
    data text
);

--
-- tables
--

select londiste.subscriber_add_table('pqueue', 'public.testdata_nopk');
select londiste.subscriber_add_table('pqueue', 'public.testdata');

select pgq.create_queue('pqueue');
select londiste.subscriber_add_table('pqueue', 'public.testdata');
select londiste.subscriber_add_table('pqueue', 'public.testdata');

select * from londiste.subscriber_get_table_list('pqueue');

select londiste.subscriber_remove_table('pqueue', 'public.nonexist');
select londiste.subscriber_remove_table('pqueue', 'public.testdata');

select * from londiste.subscriber_get_table_list('pqueue');

--
-- seqs
--

select * from londiste.subscriber_get_seq_list('pqueue');
select londiste.subscriber_add_seq('pqueue', 'public.no_seq');
select londiste.subscriber_add_seq('pqueue', 'public.testdata_id_seq');
select londiste.subscriber_add_seq('pqueue', 'public.testdata_id_seq');
select * from londiste.subscriber_get_seq_list('pqueue');
select londiste.subscriber_remove_seq('pqueue', 'public.testdata_id_seq');
select londiste.subscriber_remove_seq('pqueue', 'public.testdata_id_seq');
select * from londiste.subscriber_get_seq_list('pqueue');

--
-- linked queue
--
select londiste.subscriber_add_table('pqueue', 'public.testdata');
insert into londiste.link (source, dest) values ('mqueue', 'pqueue');


select londiste.subscriber_add_table('pqueue', 'public.testdata');
select londiste.subscriber_remove_table('pqueue', 'public.testdata');

select londiste.subscriber_add_seq('pqueue', 'public.testdata_id_seq');
select londiste.subscriber_remove_seq('pqueue', 'public.testdata_seq');

--
-- skip-truncate, set_table_state
--

select londiste.subscriber_add_table('pqueue', 'public.skiptest');
select skip_truncate from londiste.subscriber_table where table_name = 'public.skiptest';
select londiste.subscriber_set_skip_truncate('pqueue', 'public.skiptest', true);
select skip_truncate from londiste.subscriber_table where table_name = 'public.skiptest';
select londiste.subscriber_set_table_state('pqueue', 'public.skiptest', 'snap1', 'in-copy');
select skip_truncate, snapshot from londiste.subscriber_table where table_name = 'public.skiptest';
select londiste.subscriber_set_table_state('pqueue', 'public.skiptest', null, 'ok');
select skip_truncate, snapshot from londiste.subscriber_table where table_name = 'public.skiptest';

--
-- test tick tracking
--
select londiste.get_last_tick('c');
select londiste.set_last_tick('c', 1);
select londiste.get_last_tick('c');
select londiste.set_last_tick('c', 2);
select londiste.get_last_tick('c');
select londiste.set_last_tick('c', NULL);
select londiste.get_last_tick('c');


-- test triggers

create table tgfk (
  id int4 primary key,
  data text
);

create table tgtest (
  id int4 primary key,
  fk int4 references tgfk,
  data text
);

create or replace function notg() returns trigger as $$
begin
    return null;
end;
$$ language plpgsql;

create trigger tg_nop after insert on tgtest for each row execute procedure notg();

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['t']);

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

select * from londiste.subscriber_restore_all_table_triggers('tgtest');

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

-- test constraint triggers

create constraint trigger tg_nop1 after insert on tgtest deferrable initially deferred for each row execute procedure notg();
create constraint trigger tg_nop2 after update on tgtest deferrable initially immediate for each row execute procedure notg();
create constraint trigger tg_nop3 after delete on tgtest not deferrable initially immediate for each row execute procedure notg();

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['t']);

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

select * from londiste.subscriber_restore_all_table_triggers('tgtest');

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

--

select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['c']);

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

select * from londiste.subscriber_restore_all_table_triggers('tgtest');

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

--

select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['t', 'c']);

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');

select * from londiste.subscriber_restore_all_table_triggers('tgtest');

select * from londiste.find_table_triggers('tgtest');
select * from londiste.subscriber_get_table_pending_triggers('tgtest');
