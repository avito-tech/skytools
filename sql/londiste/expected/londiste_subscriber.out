set client_min_messages = 'warning';
\set VERBOSITY 'terse'
create table testdata (
    id serial primary key,
    data text
);
--
-- tables
--
select londiste.subscriber_add_table('pqueue', 'public.testdata_nopk');
 subscriber_add_table 
----------------------
                    0
(1 row)

select londiste.subscriber_add_table('pqueue', 'public.testdata');
 subscriber_add_table 
----------------------
                    0
(1 row)

select pgq.create_queue('pqueue');
 create_queue 
--------------
            1
(1 row)

select londiste.subscriber_add_table('pqueue', 'public.testdata');
ERROR:  duplicate key value violates unique constraint "subscriber_table_pkey"
select londiste.subscriber_add_table('pqueue', 'public.testdata');
ERROR:  duplicate key value violates unique constraint "subscriber_table_pkey"
select * from londiste.subscriber_get_table_list('pqueue');
      table_name      | merge_state | snapshot | trigger_name | skip_truncate 
----------------------+-------------+----------+--------------+---------------
 public.testdata_nopk |             |          |              | 
 public.testdata      |             |          |              | 
(2 rows)

select londiste.subscriber_remove_table('pqueue', 'public.nonexist');
ERROR:  no such table
select londiste.subscriber_remove_table('pqueue', 'public.testdata');
 subscriber_remove_table 
-------------------------
                       0
(1 row)

select * from londiste.subscriber_get_table_list('pqueue');
      table_name      | merge_state | snapshot | trigger_name | skip_truncate 
----------------------+-------------+----------+--------------+---------------
 public.testdata_nopk |             |          |              | 
(1 row)

--
-- seqs
--
select * from londiste.subscriber_get_seq_list('pqueue');
 subscriber_get_seq_list 
-------------------------
(0 rows)

select londiste.subscriber_add_seq('pqueue', 'public.no_seq');
 subscriber_add_seq 
--------------------
                  0
(1 row)

select londiste.subscriber_add_seq('pqueue', 'public.testdata_id_seq');
 subscriber_add_seq 
--------------------
                  0
(1 row)

select londiste.subscriber_add_seq('pqueue', 'public.testdata_id_seq');
ERROR:  duplicate key value violates unique constraint "subscriber_seq_pkey"
select * from londiste.subscriber_get_seq_list('pqueue');
 subscriber_get_seq_list 
-------------------------
 public.no_seq
 public.testdata_id_seq
(2 rows)

select londiste.subscriber_remove_seq('pqueue', 'public.testdata_id_seq');
 subscriber_remove_seq 
-----------------------
                     0
(1 row)

select londiste.subscriber_remove_seq('pqueue', 'public.testdata_id_seq');
ERROR:  no such seq?
select * from londiste.subscriber_get_seq_list('pqueue');
 subscriber_get_seq_list 
-------------------------
 public.no_seq
(1 row)

--
-- linked queue
--
select londiste.subscriber_add_table('pqueue', 'public.testdata');
 subscriber_add_table 
----------------------
                    0
(1 row)

insert into londiste.link (source, dest) values ('mqueue', 'pqueue');
select londiste.subscriber_add_table('pqueue', 'public.testdata');
ERROR:  duplicate key value violates unique constraint "subscriber_table_pkey"
select londiste.subscriber_remove_table('pqueue', 'public.testdata');
 subscriber_remove_table 
-------------------------
                       0
(1 row)

select londiste.subscriber_add_seq('pqueue', 'public.testdata_id_seq');
 subscriber_add_seq 
--------------------
                  0
(1 row)

select londiste.subscriber_remove_seq('pqueue', 'public.testdata_seq');
ERROR:  no such seq?
--
-- skip-truncate, set_table_state
--
select londiste.subscriber_add_table('pqueue', 'public.skiptest');
 subscriber_add_table 
----------------------
                    0
(1 row)

select skip_truncate from londiste.subscriber_table where table_name = 'public.skiptest';
 skip_truncate 
---------------
 
(1 row)

select londiste.subscriber_set_skip_truncate('pqueue', 'public.skiptest', true);
 subscriber_set_skip_truncate 
------------------------------
                            1
(1 row)

select skip_truncate from londiste.subscriber_table where table_name = 'public.skiptest';
 skip_truncate 
---------------
 t
(1 row)

select londiste.subscriber_set_table_state('pqueue', 'public.skiptest', 'snap1', 'in-copy');
 subscriber_set_table_state 
----------------------------
                          1
(1 row)

select skip_truncate, snapshot from londiste.subscriber_table where table_name = 'public.skiptest';
 skip_truncate | snapshot 
---------------+----------
 t             | snap1
(1 row)

select londiste.subscriber_set_table_state('pqueue', 'public.skiptest', null, 'ok');
 subscriber_set_table_state 
----------------------------
                          1
(1 row)

select skip_truncate, snapshot from londiste.subscriber_table where table_name = 'public.skiptest';
 skip_truncate | snapshot 
---------------+----------
               | 
(1 row)

--
-- test tick tracking
--
select londiste.get_last_tick('c');
 get_last_tick 
---------------
              
(1 row)

select londiste.set_last_tick('c', 1);
 set_last_tick 
---------------
             1
(1 row)

select londiste.get_last_tick('c');
 get_last_tick 
---------------
             1
(1 row)

select londiste.set_last_tick('c', 2);
 set_last_tick 
---------------
             1
(1 row)

select londiste.get_last_tick('c');
 get_last_tick 
---------------
             2
(1 row)

select londiste.set_last_tick('c', NULL);
 set_last_tick 
---------------
             1
(1 row)

select londiste.get_last_tick('c');
 get_last_tick 
---------------
              
(1 row)

-- test triggers
create table tgfk (
  id int4 primary key,
  data text
);
create table tgtest (
  id int4 primary key,
  fk int4 references tgfk,
  data text
);
create or replace function notg() returns trigger as $$
begin
    return null;
end;
$$ language plpgsql;
create trigger tg_nop after insert on tgtest for each row execute procedure notg();
select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                    trigger_def                                     | trigger_type 
---------------+--------------+------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg() | t
(1 row)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['t']);
 subscriber_drop_all_table_triggers 
------------------------------------
                                  1
(1 row)

select * from londiste.find_table_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name |                                    trigger_def                                     | trigger_type 
------------+--------------+------------------------------------------------------------------------------------+--------------
 tgtest     | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg() | t
(1 row)

select * from londiste.subscriber_restore_all_table_triggers('tgtest');
 subscriber_restore_all_table_triggers 
---------------------------------------
                                     1
(1 row)

select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                    trigger_def                                     | trigger_type 
---------------+--------------+------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg() | t
(1 row)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

-- test constraint triggers
create constraint trigger tg_nop1 after insert on tgtest deferrable initially deferred for each row execute procedure notg();
create constraint trigger tg_nop2 after update on tgtest deferrable initially immediate for each row execute procedure notg();
create constraint trigger tg_nop3 after delete on tgtest not deferrable initially immediate for each row execute procedure notg();
select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                                            trigger_def                                                            | trigger_type 
---------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg()                                                | t
 public.tgtest | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 public.tgtest | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 public.tgtest | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
(4 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['t']);
 subscriber_drop_all_table_triggers 
------------------------------------
                                  1
(1 row)

select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                                            trigger_def                                                            | trigger_type 
---------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 public.tgtest | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 public.tgtest | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
(3 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name |                                    trigger_def                                     | trigger_type 
------------+--------------+------------------------------------------------------------------------------------+--------------
 tgtest     | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg() | t
(1 row)

select * from londiste.subscriber_restore_all_table_triggers('tgtest');
 subscriber_restore_all_table_triggers 
---------------------------------------
                                     1
(1 row)

select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                                            trigger_def                                                            | trigger_type 
---------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 public.tgtest | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 public.tgtest | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg()                                                | t
(4 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

--
select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['c']);
 subscriber_drop_all_table_triggers 
------------------------------------
                                  1
(1 row)

select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                    trigger_def                                     | trigger_type 
---------------+--------------+------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg() | t
(1 row)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name |                                                            trigger_def                                                            | trigger_type 
------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 tgtest     | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 tgtest     | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 tgtest     | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
(3 rows)

select * from londiste.subscriber_restore_all_table_triggers('tgtest');
 subscriber_restore_all_table_triggers 
---------------------------------------
                                     1
(1 row)

select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                                            trigger_def                                                            | trigger_type 
---------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg()                                                | t
 public.tgtest | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 public.tgtest | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 public.tgtest | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
(4 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

--
select * from londiste.subscriber_drop_all_table_triggers('tgtest', array['t', 'c']);
 subscriber_drop_all_table_triggers 
------------------------------------
                                  1
(1 row)

select * from londiste.find_table_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name |                                                            trigger_def                                                            | trigger_type 
------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 tgtest     | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg()                                                | t
 tgtest     | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 tgtest     | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 tgtest     | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
(4 rows)

select * from londiste.subscriber_restore_all_table_triggers('tgtest');
 subscriber_restore_all_table_triggers 
---------------------------------------
                                     1
(1 row)

select * from londiste.find_table_triggers('tgtest');
  table_name   | trigger_name |                                                            trigger_def                                                            | trigger_type 
---------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------+--------------
 public.tgtest | tg_nop       | CREATE TRIGGER tg_nop AFTER INSERT ON tgtest FOR EACH ROW EXECUTE PROCEDURE notg()                                                | t
 public.tgtest | tg_nop1      | CREATE CONSTRAINT TRIGGER tg_nop1 AFTER INSERT ON tgtest DEFERRABLE INITIALLY DEFERRED FOR EACH ROW EXECUTE PROCEDURE notg()      | c
 public.tgtest | tg_nop2      | CREATE CONSTRAINT TRIGGER tg_nop2 AFTER UPDATE ON tgtest DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg()     | c
 public.tgtest | tg_nop3      | CREATE CONSTRAINT TRIGGER tg_nop3 AFTER DELETE ON tgtest NOT DEFERRABLE INITIALLY IMMEDIATE FOR EACH ROW EXECUTE PROCEDURE notg() | c
(4 rows)

select * from londiste.subscriber_get_table_pending_triggers('tgtest');
 table_name | trigger_name | trigger_def | trigger_type 
------------+--------------+-------------+--------------
(0 rows)

